File: reference-implementations/python/ledger/closing.py
"""
Closing / Retained Earnings for Python ref (mirror of src/closing/closing.ts).

Uses trial balance style to close actual Income/Expense accounts to RE.
"""

from typing import List, Optional
from .money import Money
from .account import Account, AccountType
from .journal import create_balanced_entry, JournalEntry
from .ledger import Ledger

def create_retained_earnings_account(code: str = '3100', name: str = 'Retained Earnings') -> Account:
    return Account(code, name, AccountType.Equity)

def generate_closing_entries(ledger: Ledger, close_date: str,
                             retained_earnings: Optional[Account] = None) -> List[JournalEntry]:
    re = retained_earnings or create_retained_earnings_account()
    entries: List[JournalEntry] = []
    # Use trial-like: walk entries to find temp accounts (simplified for ref)
    temp_bals: dict = {}
    for e in ledger.entries:
        for line in e.lines:
            if line.account.type in (AccountType.Income, AccountType.Expense):
                key = (line.account.code, line.account.name, line.account.type, line.amount.currency)
                if key not in temp_bals:
                    temp_bals[key] = Money.zero(line.amount.currency)
                if line.side == 'debit':
                    temp_bals[key] = temp_bals[key].sub(line.amount)  # simplistic net
                else:
                    temp_bals[key] = temp_bals[key].add(line.amount)

    seq = 0
    for (code, name, typ, curr), bal in temp_bals.items():
        if bal.is_zero():
            continue
        temp_acct = Account(code, name, typ)
        if typ == AccountType.Income:
            # Dr income, Cr RE
            entries.append(create_balanced_entry(
                f'close-{code}-{close_date}-{seq}', close_date,
                temp_acct, re, bal,
                f'Close {name} to RE', ['gaap-closing-entries-re']
            ))
        else:
            # Dr RE, Cr expense
            entries.append(create_balanced_entry(
                f'close-{code}-{close_date}-{seq}', close_date,
                re, temp_acct, bal,
                f'Close {name} to RE', ['gaap-closing-entries-re']
            ))
        seq += 1
    return entries

File: src/closing/closing.ts
import { Account, AccountType } from '../core/account.js';
import { createBalancedEntry, JournalEntry } from '../core/journal.js';
import { Ledger } from '../core/ledger.js';
import { Money } from '../core/money.js';

/**
 * Retained earnings helper account factory (typical equity account).
 * Caller may supply their own chart code/name.
 */
export function createRetainedEarningsAccount(code = '3100', name = 'Retained Earnings'): Account {
  return new Account(code, name, AccountType.Equity);
}

/**
 * Generate balanced closing JournalEntry(s) that zero temporary accounts
 * (Income/Expense) into Retained Earnings for the given closeDate.
 *
 * Uses only public Ledger APIs (trialBalance + summarizeByType grouping) so
 * multi-currency ledgers are handled correctly. Never relies on the private
 * pickCurrency used by the minimal incomeStatement view.
 *
 * Every returned entry is created via createBalancedEntry so it is guaranteed
 * to pass validateEntry and preserve the fundamental equation when applied.
 *
 * Citations: attach matching principle + period closing canon.
 */
export function generateClosingEntries(
  ledger: Ledger,
  closeDate: string,
  retainedEarnings: Account = createRetainedEarningsAccount()
): JournalEntry[] {
  const entries: JournalEntry[] = [];
  const sums = ledger.summarizeByType();

  // Group by currency explicitly (trialBalance already gives one row per curr)
  const byCurrency = new Map<string, { income: Money; expense: Money }>();

  for (const s of sums) {
    if (!byCurrency.has(s.total.currency)) {
      byCurrency.set(s.total.currency, {
        income: Money.zero(s.total.currency),
        expense: Money.zero(s.total.currency),
      });
    }
    const g = byCurrency.get(s.total.currency)!;
    if (s.type === AccountType.Income) g.income = g.income.add(s.total);
    if (s.type === AccountType.Expense) g.expense = g.expense.add(s.total);
  }

  // Premium: discover the actual temporary accounts from the ledger and close each individually.
  // This zeros the real Income/Expense accounts (much better than synthetic "summary").
  const tempBalances = ledger.trialBalance()
    .filter(r => r.account.type === AccountType.Income || r.account.type === AccountType.Expense)
    .filter(r => !r.balance.isZero());

  let seq = 0;
  for (const { account: tempAcct, balance } of tempBalances) {
    if (tempAcct.type === AccountType.Income) {
      // Income has credit balance normally; to close: Dr Income, Cr RE
      entries.push(
        createBalancedEntry(
          `close-${tempAcct.code}-${closeDate}-${seq++}`,
          closeDate,
          tempAcct,           // debit the income account (zeros its credit bal)
          retainedEarnings,   // credit RE
          balance,
          `Close ${tempAcct.name} to retained earnings`,
          ['gaap-closing-entries-re', 'gaap-matching-principle-detail']
        )
      );
    } else {
      // Expense has debit balance; to close: Dr RE, Cr Expense
      entries.push(
        createBalancedEntry(
          `close-${tempAcct.code}-${closeDate}-${seq++}`,
          closeDate,
          retainedEarnings,
          tempAcct,           // credit the expense account (zeros its debit bal)
          balance,
          `Close ${tempAcct.name} to retained earnings`,
          ['gaap-closing-entries-re', 'gaap-matching-principle-detail']
        )
      );
    }
  }

  return entries;
}
